aws/s3/helpers.py

```diff
@@ -5,7 +5,31 @@ def get_s3_bucket_name(bucket):
     return get_param_id(bucket, "Name")
 
 
-def get_s3_bucket_name_only(bucket):
-    if isinstance(bucket, dict) and "Name" in bucket:
-        return get_s3_bucket_name(bucket)
+def get_s3_resource_id(resource):
+    if isinstance(resource, dict) and "Name" in resource:
+        return get_s3_bucket_name(resource)
```

Comment:
This structure feels wrong to me -- shouldn't it return a concatenation of all the matching key's values?

I.e. is it guaranteed (or a requirement) that "resource" have at most one of the keys of interest?  If so, I'd love to see that as a comment.

Reply:
No, as the test id serves two functions:
* A unique identifier for the resource(s) being tested
* A best effort name for the resource(s) in question

For certain tests, the test id will work as all of the additional data you need to know, i.e. the test id for MFA tests being the username. But for a number of other tests, i.e. security group / firewall tests, it's just there to point you to the security group in question which you will then have to go and look at via the AWS/GCP tooling. Or is some other cases, you will just need to look at the metadata associated with the test result.

If a resource has multiple matching keys, the ordering is in place to attempt to get the most representative identifier for the resource.

Reply:
Ahh, if there is a "magical order" to the statements, I _really_ think that needs to be documented in the code. Otherwise it's extremely brittle for any maintainer who doesn't grok the full service ontology.

Especially since I could see this code being an "obvious" modification point for the "file name too long" issue.

Reply:
There isn't a magic order, but nevertheless I do agree that this current way of handling test ids is extremely brittle.

I'm beginning to think that it might make sense to bring these "get resource id" functions closer to the calls within `resources.py`. Maybe the current `resources.py` functions should be turned into classes with a "get_resources" and a "get_id" function required for each, with some mixins that can be used for common ids. Would additionally require a refactor around how we use `parameterize`, but I think that would be worth it.

Reply:
> I'm beginning to think that it might make sense to bring these "get resource id" functions closer to the calls within resources.py. Maybe the current resources.py functions should be turned into classes with a "get_resources" and a "get_id" function required for each, with some mixins that can be used for common ids. Would additionally require a refactor around how we use parameterize, but I think that would be worth it.

+1 dataclasses or a validator library could help too.

At some point I think it makes sense to switch to boto3 since it looks like it provides and ID function https://boto3.amazonaws.com/v1/documentation/api/latest/reference/core/resources.html#boto3.resources.model.Identifier (it'd also be more familiar to Ops and provide more pythonic attribute access)

Reply:
fwiw - see also my approach at:
 - [usage](https://github.com/hwine/frost/blob/branch_check/github/branches/test_branch_protection.py#L27-L29)
- definition [in dataclass](https://github.com/hwine/frost/blob/branch_check/github/branches/validate_compliance.py#L19-L24)

I'm not sure a mixin is justified -- we're simply adding one method. Mixins come with complexity and future maintenance issues.

Reply:
I've combined this discussion into a new issue: https://github.com/mozilla/frost/issues/418

@hwine Are you ok with us merging this as-is and keeping the above discussion as a follow-up issue?

Reply:
go for it